sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""This module contains Splittable DoFn logic that is specific to DirectRunner.
+"""
+
+from threading import Lock
+from threading import Timer
+
+import apache_beam as beam
+from apache_beam import TimeDomain
+from apache_beam import pvalue
+from apache_beam.io.iobase import RestrictionTracker
+from apache_beam.pipeline import PTransformOverride
+from apache_beam.runners.common import DoFnContext
+from apache_beam.runners.common import DoFnInvoker
+from apache_beam.runners.common import DoFnSignature
+from apache_beam.runners.common import OutputProcessor
+from apache_beam.runners.direct.evaluation_context import DirectStepContext
+from apache_beam.runners.direct.util import KeyedWorkItem
+from apache_beam.runners.direct.watermark_manager import WatermarkManager
+from apache_beam.runners.sdf_common import ElementAndRestriction
+from apache_beam.runners.sdf_common import ProcessKeyedElements
+from apache_beam.transforms.core import ProcessContinuation
+from apache_beam.transforms.ptransform import PTransform
+from apache_beam.transforms.trigger import _ValueStateTag
+from apache_beam.utils.windowed_value import WindowedValue
+
+
+class ProcessKeyedElementsViaKeyedWorkItemsOverride(PTransformOverride):
+  """A transform override for ProcessElements transform."""
+
+  def get_matcher(self):
+    def _matcher(applied_ptransform):
+      return isinstance(
+          applied_ptransform.transform, ProcessKeyedElements)
+
+    return _matcher
+
+  def get_replacement_transform(self, ptransform):
+    return ProcessKeyedElementsViaKeyedWorkItems(ptransform)
+
+
+class ProcessKeyedElementsViaKeyedWorkItems(PTransform):
+  """A transform that processes Splittable DoFn input via KeyedWorkItems.
+  """
+
+  def __init__(self, process_keyed_elements_transform):
+    self._process_keyed_elements_transform = process_keyed_elements_transform
+
+  def expand(self, pcoll):
+    return pcoll | beam.core.GroupByKey() | ProcessElements(
+        self._process_keyed_elements_transform)
+
+
+class ProcessElements(PTransform):
+  """Processes keyed input via Splittable DoFn objects."""
+
+  def __init__(self, process_keyed_elements_transform):
+    self._process_keyed_elements_transform = process_keyed_elements_transform
+    self.sdf = self._process_keyed_elements_transform.sdf
+
+  def expand(self, pcoll):
+    return pvalue.PCollection(pcoll.pipeline)
+
+  def new_process_fn(self, sdf):
+    return ProcessFn(
+        sdf,
+        self._process_keyed_elements_transform.ptransform_args,
+        self._process_keyed_elements_transform.ptransform_kwargs)
+
+
+class ProcessFn(beam.DoFn):
+
+  def __init__(
+      self, sdf, args_for_invoker, kwargs_for_invoker):
+    self.sdf = sdf
+    self._element_tag = _ValueStateTag('element')
+    self._restriction_tag = _ValueStateTag('restriction')
+    self.watermark_hold_tag = _ValueStateTag('watermark_hold')
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
Why is this the only "_tag" not beginning with an underscore?